webapp/src/main/java/io/github/microcks/service/SpanStorageService.java

```diff
@@ -81,26 +82,40 @@ public SpanStorageService(ApplicationEventPublisher publisher) {
     */
    public void storeSpan(ReadableSpan span) {
       String traceId = span.getSpanContext().getTraceId();
-
-      // Add the span to our collection
-      spansByTraceId.computeIfAbsent(traceId, k -> new ArrayList<>()).add(span);
-
-      // Limit spans per trace to prevent memory issues
-      List<ReadableSpan> spans = spansByTraceId.get(traceId);
-      if (spans.size() > MAX_SPANS_PER_TRACE) {
-         // Remove oldest span to cap memory
-         spans.removeFirst();
+      SpanData spanData = span.toSpanData();
+
+      // Add the span to our collection. Use a synchronized list for per-trace spans so concurrent
+      // additions/removals are thread-safe. Also synchronize on the list when doing size+remove
+      // operations to make them atomic.
+      List<SpanData> spans = spansByTraceId.computeIfAbsent(traceId,
+            k -> Collections.synchronizedList(new ArrayList<>()));
+
+      synchronized (spans) {
+         spans.add(new SpanDataDTO(spanData));
+
+         // Limit spans per trace to prevent memory issues
+         if (spans.size() > MAX_SPANS_PER_TRACE) {
+            // Remove oldest span to cap memory
+            spans.remove(0);
+         }
       }
 
-      // Limit total number of traces to prevent memory leaks
-      if (spansByTraceId.size() > MAX_TRACES) {
-         // Remove oldest trace this works because LinkedHashMap maintains insertion order
-         String oldestTraceId = spansByTraceId.keySet().iterator().next();
-         spansByTraceId.remove(oldestTraceId);
+      // Limit total number of traces to prevent memory leaks. Iteration over the
+      // synchronizedMap's keySet must be done while holding the map lock to avoid
+      // ConcurrentModificationException.
+      synchronized (spansByTraceId) {
+         if (spansByTraceId.size() > MAX_TRACES) {
+            // Remove oldest trace - LinkedHashMap preserves insertion order
+            var it = spansByTraceId.keySet().iterator();
+            if (it.hasNext()) {
+               it.next();
+               it.remove();
+            }
```

Comment:
I would have keep something like:
```java
// Limit total number of traces to prevent memory leaks
      if (spansByTraceId.size() > MAX_TRACES) {
         // Remove oldest trace this works because LinkedHashMap maintains insertion order
         synchronized (spansByTraceId) {
            String oldestTraceId = spansByTraceId.keySet().iterator().next();
            spansByTraceId.remove(oldestTraceId);
         }
      }
```